lib/active_merchant/billing/gateways/firstdata_e4_v27.rb

```diff
@@ -0,0 +1,445 @@
+          :month      => params[4],
+          :year       => params[5])
+
+        xml.tag! 'TransarmorToken', params[0]
+        xml.tag! 'Expiry_Date', expdate(credit_card)
+        xml.tag! 'CardHoldersName', credit_card.name
+        xml.tag! 'CardType', card_type(credit_card.brand)
+        add_card_authentication_data(xml, options)
+      end
+
+      def add_customer_data(xml, options)
+        xml.tag! 'Customer_Ref', options[:customer] if options[:customer]
+        xml.tag! 'Client_IP', options[:ip] if options[:ip]
+        xml.tag! 'Client_Email', options[:email] if options[:email]
+      end
+
+      def add_address(xml, options)
+        if address = options[:billing_address] || options[:address]
+          xml.tag! 'Address' do
+            xml.tag! 'Address1', address[:address1]
+            xml.tag! 'Address2', address[:address2] if address[:address2]
+            xml.tag! 'City', address[:city]
+            xml.tag! 'State', address[:state]
+            xml.tag! 'Zip', address[:zip]
+            xml.tag! 'CountryCode', address[:country]
+          end
+          xml.tag! 'ZipCode', address[:zip]
+        end
+      end
+
+      def add_invoice(xml, options)
+        xml.tag! 'Reference_No', options[:order_id]
+        xml.tag! 'Reference_3',  options[:description] if options[:description]
+      end
+
+      def add_tax_fields(xml, options)
+        xml.tag! 'Tax1Amount',  options[:tax1_amount] if options[:tax1_amount]
+        xml.tag! 'Tax1Number',  options[:tax1_number] if options[:tax1_number]
+      end
+
+      def add_level_3(xml, options)
+        xml.tag!('Level3') { |x| x << options[:level_3] } if options[:level_3]
+      end
+
+      def expdate(credit_card)
+        "#{format(credit_card.month, :two_digits)}#{format(credit_card.year, :two_digits)}"
+      end
+
+      def card_type(credit_card_brand)
+        E4_BRANDS[credit_card_brand.to_sym] if credit_card_brand
+      end
+
+      def commit(action, data, credit_card = nil)
+        url = (test? ? self.test_url : self.live_url)
+        request = build_request(action, data)
+        begin
+          response = parse(ssl_post(url, request, headers('POST', url, request)))
+        rescue ResponseError => e
+          response = parse_error(e.response)
+        end
+
+        Response.new(successful?(response), message_from(response), response,
+          :test => test?,
+          :authorization => successful?(response) ? response_authorization(action, response, credit_card) : '',
+          :avs_result => {:code => response[:avs]},
+          :cvv_result => response[:cvv2],
+          :error_code => standard_error_code(response)
+        )
+      end
+
+      def headers(method, url, request)
+        content_type = 'application/xml'
+        content_digest = Digest::SHA1.hexdigest(request)
+        sending_time = Time.now.utc.iso8601
+        payload = method + "\n" + content_type + "\n" + content_digest + "\n" + sending_time + "\n" + url.split('.com')[1]
```

Comment:
This is nitpicky, but `payload = [method, content_type, content_digest, sending_time, url.split('.com')[1]].join("\n")` might be easier to read.